### PR TITLE
Sns sender

### PR DESCRIPTION
**Pull request by devinsba**

### CODE REVIEW

File: reporter-sns/src/main/java/zipkin/reporter/sns/SnsSender.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        publishRequest.addMessageAttributesEntry(
+                "spans",
+                new MessageAttributeValue()
+                        .withDataType("Binary." + encoding().name())
+                        .withBinaryValue(ByteBuffer.wrap(BytesMessageEncoder.forEncoding(encoding()).encode(list))));
+        get().publish(publishRequest);
+    }
+
+    public CheckResult check() {
+        return CheckResult.OK;
```
Comment (devinsba) [inline reporter-sns/src/main/java/zipkin/reporter/sns/SnsSender.java]: Obviously need something here

**Comment by adriancole** [comment]
looks like a good start. is there any way to have amazon's client point at a local url for testing? if so, we could use mockwebserver.

**Comment by devinsba** [comment]
Looks like all the AS clients have a setEndpoint method:

http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/AmazonWebServiceClient.html#setEndpoint-java.lang.String-

**Comment by llinder** [comment]
We should add a style guide.  I have been following: https://github.com/openzipkin/zipkin-reporter-java/blob/master/CONTRIBUTING.md but don't really have a preference.  Mostly I just want to make sure my PR is inline with yours.

File: reporter-sns/src/main/java/zipkin/reporter/sns/SnsSender.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+@AutoValue
+public abstract class SnsSender extends LazyCloseable<AmazonSNSClient> implements Sender {
+    abstract String topicArn();
+    abstract AWSCredentialsProvider awsCredentialsProvider();
+    public abstract Encoding encoding();
+
+    public static Builder builder() {
+        return new AutoValue_SnsSender.Builder()
+                .awsCredentialsProvider(new DefaultAWSCredentialsProviderChain())
+                .encoding(Encoding.JSON);
```
Comment (llinder) [inline reporter-sns/src/main/java/zipkin/reporter/sns/SnsSender.java]: Is your preference to default to JSON?  If so I will switch the SQS default to match yours.

**Comment by adriancole** [comment]
I'd default them to thrift, json is really only helpful in debugging, as
you can stuff more spans in a message when not json, also the overhead of
encoding is half.

File: reporter-sns/src/main/java/zipkin/reporter/sns/SnsSender.java (hunk 2 above)
Comment (adriancole) [inline reporter-sns/src/main/java/zipkin/reporter/sns/SnsSender.java]: I think there's an attribute to enable "raw" (non-json) messages http://docs.aws.amazon.com/sns/latest/dg/large-payload-raw-message.html

opened on related issues about encoding https://github.com/openzipkin/zipkin-aws/issues/15

**Comment by devinsba** [comment]
I'll adjust this to match what is in the SQS sender once that is more finalized and try to pull out some common code between the 2.

**Comment by devinsba** [comment]
I totally forgot about this. Hopefully I can get this finished up soon. I don't know if we are going to end up using it though. Kinesis is probably more likely.

**Comment by adriancole** [comment]
Well if theres no one planning to use it, closing is probably fine..

**Comment by devinsba** [comment]
Closing for now, will revisit if we end up reconsidering this approach

**Pull request closed without merging**